posthog/tasks/alerts/checks.py

```diff
@@ -18,21 +19,146 @@
+from posthog.schema import TrendsQuery, IntervalType, ChartDisplayType, NodeKind, AlertCalculationInterval, AlertState
+from posthog.utils import get_from_dict_or_attr
+from posthog.caching.fetch_from_cache import InsightResult
+from posthog.clickhouse.client.limit import limit_concurrency
+from prometheus_client import Gauge
+from django.db.models import Q
+
+
+HOURLY_ALERTS_BACKLOG_GAUGE = Gauge(
+    "hourly_alerts_backlog",
+    "Number of hourly alerts that are not being checked in the last hour.",
+)
+
+DAILY_ALERTS_BACKLOG_GAUGE = Gauge(
+    "daily_alerts_backlog",
+    "Number of daily alerts that are not being checked in the last 24 hours.",
+)
 
 logger = structlog.get_logger(__name__)
 
+WRAPPER_NODE_KINDS = [NodeKind.DATA_TABLE_NODE, NodeKind.DATA_VISUALIZATION_NODE, NodeKind.INSIGHT_VIZ_NODE]
+
+NON_TIME_SERIES_DISPLAY_TYPES = {
+    ChartDisplayType.BOLD_NUMBER,
+    ChartDisplayType.ACTIONS_PIE,
+    ChartDisplayType.ACTIONS_BAR_VALUE,
+    ChartDisplayType.ACTIONS_TABLE,
+    ChartDisplayType.WORLD_MAP,
+}
+
+
+@shared_task(
+    ignore_result=True,
+    expires=60 * 60,
+)
+def hourly_alerts_backlog_task() -> None:
+    """
+    This runs every 5min to check for alerts with hourly check SLA
+    but haven't been checked in the last hour
+    """
+    now = datetime.now(UTC)
+
+    alerts_breaching_sla = AlertConfiguration.objects.filter(
+        Q(
+            enabled=True,
+            calculation_interval=AlertCalculationInterval.HOURLY,
+            last_checked_at__lte=now - relativedelta(hours=1),
+        )
+    ).count()
+
+    HOURLY_ALERTS_BACKLOG_GAUGE.set(alerts_breaching_sla)
+
+
+@shared_task(
+    ignore_result=True,
+    expires=60 * 60,
+)
+def daily_alerts_backlog_task() -> None:
+    """
+    This runs every 15min to check for alerts with hourly check SLA
+    but haven't been checked in the last hour
+    """
+    now = datetime.now(UTC)
+
+    alerts_breaching_sla = AlertConfiguration.objects.filter(
+        Q(
+            enabled=True,
+            calculation_interval=AlertCalculationInterval.HOURLY,
+            last_checked_at__lte=now - relativedelta(days=1),
+        )
+    ).count()
+
+    DAILY_ALERTS_BACKLOG_GAUGE.set(alerts_breaching_sla)
+
+
+@shared_task(
+    ignore_result=True,
+    expires=60 * 60,
+)
+def hourly_alerts_task() -> None:
+    """
+    This runs every 5min to check for alerts that have 5min left to recalculate
+    """
+    now = datetime.now(UTC)
+    # alerts with 5min left for recalculation
+    last_checked_at_before = now - relativedelta(minutes=55)
+    check_all_alerts(AlertCalculationInterval.HOURLY, last_checked_at_before)
+
+
+@shared_task(
+    ignore_result=True,
+    expires=60 * 60,
+)
+def daily_alerts_task() -> None:
+    """
+    This runs every hour to check for alerts that have 5min left to recalculate
+    """
+    now = datetime.now(UTC)
+    # alerts with 1h left for recalculation
+    last_checked_at_before = now - relativedelta(hours=23)
```

Comment:
Can alerts be shifted this way? For example, I created an alert at 1 PM, so each matched condition will move it one hour behind. After eight alerts, it will invocate at 9 PM.